Standardize on ej:valid validation

christophermaier opened the pull request:
Adds necessary `#ej_invalid{}` matching clauses to `malformed_message_request/3`.  Standardizes many error messages, deleting much of the duplicated and duplicated-except-for-some-tiny-differences code that was scattered across multiple endpoints.

These PRs are required for full effect:
- https://github.com/opscode/chef_objects/pull/17
- https://github.com/opscode/chef-pedant-core/pull/10
- https://github.com/opscode/chef-pedant-tests/pull/8

Commit:
Remove unused malformed_request_message/3

Commit:
Pull many ej_invalid messages up to chef_wm_base

Commit:
Delete redundant ej_invalid message handlers

Commit:
Add ej_invalid handler for sandboxes

seth commented on src/chef_wm_base.erl:
```diff
@@ -191,7 +192,7 @@ malformed_request_message({client_name_mismatch}, _Req, _State) ->
 malformed_request_message({bad_client_name, Name, Pattern}, _Req, _State) ->
     {[{<<"error">>, [iolist_to_binary(["Invalid client name '", Name,
-				       "' using regex: '", Pattern, "'."])]}]};
+                                       "' using regex: '", Pattern, "'."])]}]};
```
Personally, I find the "here's the regex we used" error message rather too exposing of implementation detail. But I'm open to other opinions as well as addressing a fix in a different (later) PR.

seth replied:
So since below we are going generic for invalid field, maybe we should just do the same here?

seth commented on src/chef_wm_base.erl:
```diff
@@ -207,9 +208,109 @@ malformed_request_message({bad_run_lists, {Field, _Value}}, _Req, _State) ->
+malformed_request_message(#ej_invalid{type=Type,
+                                      key=Key,
+                                      msg=Expected
+                                     }, _Req, _State) when Type =:= exact orelse
```
orelse is fine, but a ';' would work as well. When describing alternatives for the same argument, I think ';' reads better. When different arguments are in play, I've seen weird behvaior and found orelse/andalso better

seth commented on src/chef_wm_base.erl:
```diff
@@ -207,9 +208,109 @@ malformed_request_message({bad_run_lists, {Field, _Value}}, _Req, _State) ->
+    error_envelope([<<"Field '", Key/binary, "' invalid">>]);
+
+%% Things that should be hashes, but aren't.
+malformed_request_message(#ej_invalid{type=json_type,
```
Doesn't ej:valid give you the expected type? Wonder if that would help simplify this at all? I'm not clear why we have to match specifically on the keys here. Feels more coupled than I would like.

seth commented on src/chef_wm_base.erl:
```diff
@@ -207,9 +208,109 @@ malformed_request_message({bad_run_lists, {Field, _Value}}, _Req, _State) ->
+    error_envelope([<<"Field '", Key/binary, "' is not a hash">>]);
+
+%% Entire run list is the wrong type
+malformed_request_message(#ej_invalid{type=json_type,
```
Is this really any more clear than the generic: "invalid field 'run_list'"?

seth commented on src/chef_wm_base.erl:
```diff
@@ -207,9 +208,109 @@ malformed_request_message({bad_run_lists, {Field, _Value}}, _Req, _State) ->
+    list_to_binary(integer_to_list(I));
+to_binary(B) when is_binary(B)->
+    B;
+to_binary(O) ->
```
Why this? Seems like it would end up leaking erlang terms, which I would like to avoid in any data going out to users.

seth commented on src/chef_wm_base.erl:
```diff
@@ -48,6 +48,7 @@
 -export([check_cookbook_authz/3,
          delete_object/3]).
 
```
What would you think about moving the malformed_request_message stuff into its own module? Feels like a lot of code.

seth commented on src/chef_wm_clients.erl:
```diff
@@ -137,41 +137,5 @@ all_clients_json(Req, #base_state{chef_db_context = DbContext,
     UriMap = [ {Name, RouteFun(Name)} || Name <- ClientNames ],
     ejson:encode({UriMap}).
 
-% TODO: this could stand refactoring: I'm sure there is stuff re-used by other
```
yay, deleted code!

seth commented on src/chef_wm_sandboxes.erl:
```diff
@@ -94,10 +94,11 @@ from_json(Req, #base_state{chef_db_context = DbContext,
-    chef_wm_util:error_message_envelope(Msg);
-malformed_request_message({bad_checksum, Msg}, _Req, _State) ->
-    chef_wm_util:error_message_envelope(Msg);
+malformed_request_message(#ej_invalid{type=fun_match,
```
Not sure I'm reading closely enough, but why do we need special code here since the invalid message looks standard? Oh, because it is a fun_match, yes?

seth commented:
Overall this looks very reasonable. I haven't yet built it for testing. A few comments above on particulars of error messages. And possibly moving the bulk of the message catching into its own module.

Commit:
Use ';' instead of 'orelse'

Commit:
Extract malformed_request_message/3 into separate module

That code was getting rather large; hopefully this makes things a
little more clear.

Commit:
Fix chef_wm_clients by using chef_wm_malformed module

Somehow this module got missed and was using the helper function that
was removed from chef_wm_util.
